### ISSUE TITLE
[css-position-3]  Section 3.5.1 Conflicts with CSS2 Positioning Behavior

### ISSUE THREAD

**Issue by gitspeaks** [w3c/csswg-drafts#11258]
The rule in [Section 3.5.1](https://www.w3.org/TR/css-position-3/#resolving-insets) states:

> If only one inset property in a given axis is auto, it is set to zero.

This implies that when one inset property (e.g., left or right) is `auto` and the other is not, the `auto` value is replaced with 0. Consequently, both inset properties on that axis are treated as having non-`auto` values, with one being explicitly set to zero.

However, the CSS2 specification defines several rules that depend on scenarios where one inset property is `auto` and the other is not. For example, from [CSS 2, Section 10.3.7](https://drafts.csswg.org/css2/#abs-non-replaced-width):

Case 1: `left` and `width` are `auto`, and `right` is not `auto`; then the `width` is shrink-to-fit, and solve for `left`.
Case 3: `width` and `right` are `auto`, and `left` is not `auto`; then the `width` is shrink-to-fit, and solve for `right`.
Case 4: `left` is `auto`, and `width` and `right` are not `auto`; then solve for `left`.

By automatically setting an `auto` inset to zero when the other inset on the same axis is non-`auto`, the CSS Positioned Layout Module Level 3 invalidates these cases from CSS2 and would result in a different positioning behavior according to other CSS2  absolute position cases.

**Comment by gitspeaks** [comment]
Examples (tested in both Chrome and Firefox)

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style>
      #parent {
        position: relative;
        width: 200px;
        height: 200px;
        border: 5px solid black;
      }

      #child {
        position: absolute;
        background-color: red;
        left: 10px;
        width: 100px;
        height: 50px;
        /* Computes to 90px in compliance with CSS 2, Case 6 */
        /* https://drafts.csswg.org/css2/#abs-non-replaced-width */
        right: auto; 
        margin-left: auto;
        margin-right: auto;
      }
    </style>
  </head>
  <body>
    <div id="parent">
      <div id="child"></div>
    </div>
  </body>
</html>
```
![case6](https://github.com/user-attachments/assets/8b14f0dc-f62f-42ca-9737-3d37e300b200)

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style>
      #parent {
        position: relative;
        width: 200px;
        height: 200px;
        border: 5px solid black;
      }

      #child {
        position: absolute;
        background-color: red;
        left: 10px;
        width: 100px;
        height: 50px;
        /* 
          According to Section 3.5.1, 'auto' is converted to 0. 
          The rendered output complies with the CSS 2 specification, 
          which states: "solve the equation under the extra constraint 
          that the two margins get equal values."
          https://drafts.csswg.org/css2/#abs-non-replaced-width
        */
        right: 0;
        margin-left: auto; /* 45px */
        margin-right: auto; /* 45px */
      }
    </style>
  </head>
  <body>
    <div id="parent">
      <div id="child"></div>
    </div>
  </body>
</html>
```

![margin-case](https://github.com/user-attachments/assets/e75bb484-0a86-43bb-ba91-c3cbf5ecdaaf)

**Comment by Loirooriol** [comment]
> the CSS2 specification defines several rules

Of course these are overridden by CSS Position and CSS Align.

> would result in a different positioning behavior

Your examples seem to behave the same. https://drafts.csswg.org/css-position/#abspos-margins

 - 1st example: both `auto` margins resolve to 0 since there is an `auto` inset.
 - 2nd example: both `auto` margins resolve to `((200px - 10px - 0px) - 100px) / 2 = 45px` since there is no `auto` inset.

**Comment by Loirooriol** [comment]
OK, I get what you are saying. What the spec says "If only one inset property in a given axis is auto, it is set to zero", this is just for computing the inset-modified containing block. It's still `auto` for the purpose of resolving margins.

I guess it might be clarified a bit.

**Comment by gitspeaks** [comment]
>this is just for computing the inset-modified containing block
 
What is the expected output for the first example under CSS3 ?

```css
#child {
  position: absolute;
  background-color: red;
  left: 10px;
  width: 100px;
  height: 50px;
  /* Computes to 90px in compliance with CSS 2, Case 6 */
  /* https://drafts.csswg.org/css2/#abs-non-replaced-width */
  right: auto; 
  margin-left: auto;
  margin-right: auto;
}
```

```js
  const element = document.getElementById("child");
  const styles = getComputedStyle(element);
  console.log("Right:", styles.right);
```

**Comment by Loirooriol** [comment]
Good point. I'm not sure if it would be web compatible to change it, but even if so, it wouldn't be great if you get `0px`, assign it back, and this suddenly changes the margins.

**Comment by Loirooriol** [comment]
So I think we need the 2nd option from https://github.com/w3c/csswg-drafts/issues/11242#issuecomment-2486131051

**Comment by gitspeaks** [comment]
Second option from https://github.com/w3c/csswg-drafts/issues/11242#issuecomment-2486131051:

> Stop saying that a lone `auto` inset resolves to zero. Instead, resolve the `auto` inset to whatever value makes the size of the inset-modified containing block match the fit-content size of the abspos. This fit-content size  shrinks-to-fit into the original containing block size minus the  non-auto inset, or equivalently, into a tentative inset-modified  containing block with the `auto` offset resolved to 0.



Option Breakdown:

> Stop saying that a lone `auto` inset resolves to zero

- For CSS2 developers, the `auto` inset remains `auto`.
- For CSS3 developers, `auto` inset represents an offset of an 'inset-modified containing block' edge.


> resolve the `auto` inset to whatever value makes the size of the inset-modified containing block match the fit-content size of the abspos

CSS3 Positioning example:

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style>
      #parent {
        position: relative;
        width: 200px;
        height: 200px;
        border: 5px solid black;
      }

      #child {
        position: absolute;
        background-color: red;
        width: 50px;
        height: 50px;
        right: 0px; 
        left: 0px;
        margin-left: auto;
        margin-right: auto;
      }
    </style>
  </head>
  <body>
    <div id="parent">
      <div id="child"></div>
    </div>
  </body>
</html>
```



Available Information:

1. The `left` and `right` edges of the inset-modified containing block (blue lines below) are known.

2. As per the [spec](https://www.w3.org/TR/css-position-3/#abspos-margins):

If either inset property (`left` or `right`) on the relevant axis is `auto`, any `auto` margins resolve to `0`. Therefore, for the `#child` element, both `margin-left` and `margin-right` resolve to `0`.


Illustration:

![image](https://github.com/user-attachments/assets/8a0c6a7d-2a8d-430d-8dc6-06e98250549a)

There is no information about the exact position of the `#child` element within the `#parent`. As a result, it is impossible to determine the size of the inset-modified containing block based on any specific offsets, regardless of the fit-content size dimensions of the `#child`.

> "Two ways":
>
> 1. This fit-content size  shrinks-to-fit into the original containing block size minus the  non-auto inset,
> 2. This fit-content size  shrinks-to-fit into a tentative inset-modified  containing block with the `auto` offset resolved to 0.

Neither approach determines the exact position of the `#child` element within the `#parent`, making it impossible to establish the inset-modified containing block's offset.

**Comment by Loirooriol** [comment]
> For CSS2 developers, the auto inset remains auto.

I don't know what you mean here, CSS2 resolves `auto` too.

> CSS3 Positioning example

This is not a "lone `auto`" case. We know both insets in the inline axis, and none in the block one.

> If either inset property (left or right) on the relevant axis is auto, any auto margins resolve to 0. Therefore, for the #child element, both margin-left and margin-right resolve to 0.

Doesn't follow? As you just said, both insets are 0, so none of them is `auto`, so the margins don't resolve to 0.

> There is no information about the exact position of the #child element within the #parent

As I said several times, the position is decided by the self-alignment properties, once we know the size of the inset-modified containing block and the abspos.

> it is impossible to determine the size of the inset-modified containing block based on any specific offsets

How so? Here we know both insets so the inset-modified containing block is `200px - 0px - 0px = 200px` wide.

**Comment by gitspeaks** [comment]
>I don't know what you mean here, CSS2 resolves auto too.

For CSS2 developers, there is no change in the semantics of auto (i.e., it is not resolved to 0).

 >This is not a "lone auto" case. We know both insets in the inline axis, and none in the block one.
 
I assumed your suggestion was general and attempted to explore how it might apply to a different scenario.
 
>Doesn't follow? As you just said, both insets are 0, so none of them is auto, so the margins don't resolve to 0.

You're right—my mistake. In this case, the margins remain, and we can calculate the horizontal position.

>it is impossible to determine the size of the inset-modified containing block based on any specific offsets

>How so? Here we know both insets so the inset-modified containing block is 200px - 0px - 0px = 200px wide.

1. How do we solve for the top and bottom when auto is no longer converted to 0?
2. How is the vertical position of #child determined in this scenario?

**Comment by gitspeaks** [comment]
>This is not a "lone auto" case. We know both insets in the inline axis, and none in the block one

You're right. I deleted the posts to keep the discussion focused and relevant.

I agree that in a 'lone auto' case, 2nd option from https://github.com/w3c/csswg-drafts/issues/11242#issuecomment-2486131051 would work.

**Comment by gitspeaks** [comment]
> So I think we need the 2nd option from [#11242 (comment)](https://github.com/w3c/csswg-drafts/issues/11242#issuecomment-2486131051)

I agree.

**Comment by gitspeaks** [comment]
Not to imply any lack of appreciation for the proposed solution, I must admit I struggle to understand the value the new 'inset-modified containing block' based model brings compared to the CSS2 'containing block equation model'. 
In particular, this is unclear when considering edge cases like this and when different configurations of `auto` values still affect margin calculations. Simply put, why can't the self-alignment properties be added on top of the CSS2 model, overriding the expected rules as needed? Why is the concept of an  'inset-modified containing block' necessary to support self-alignment properties?

**Comment by gitspeaks** [comment]
The "new model" proposes the following mental approach:

1. Establish a frame in reference to the "absolute-position containing block", called the "inset-modified containing block" (IMCB).  
2. Position the absolutely positioned element within the IMCB.

Now, consider the first example in this issue using the proposed solution:

```css
#child {
    position: absolute;
    background-color: red;
    left: 10px;
    width: 100px;
    height: 50px;
    right: auto;
    margin-left: auto;
    margin-right: auto;
}
```

1. First, the `left` of the IMCB is established, and the `top` (and `bottom`) are implied by the default `self-alignment` property (`self-start`), which dictates that `top` is set according to the static position of `#child`, and `bottom` is set to `0`.

2. Second, according to the proposed solution, the bottom-right corner of the IMCB is determined based on the shrink-to-fit dimensions of `#child`, presumably resizing an already determined IMCB.

It is this second step that still requires developers to think about offsets relative to the containing block, which essentially exercises the same mental model as in CSS2.

**Comment by Loirooriol** [comment]
> I struggle to understand the value the new 'inset-modified containing block' based model brings compared to the CSS2 'containing block equation model'

The value is that, for overconstrained cases, instead of just ignoring `left` or `right`, you can use CSS align properties to align withing the inset-modified containing block.

And it's not like it's a radically different model, it's just a slightly different way of explaining the CSS2 behavior.

**Comment by gitspeaks** [comment]
> The value is that, for overconstrained cases, instead of just ignoring left or right, you can use CSS align properties to align withing the inset-modified containing block.

 I don't understand what you are saying.
 
 In [CSS2](https://www.w3.org/TR/css-position-3/#abspos-old), "over-constrained" refers to a situation where the sum of the positioned element’s properties for a given axis exceeds the size of the corresponding axis of the absolute-position containing block. 
 
>``` left + margin-left + border-left-width + padding-left + width + padding-right + border-right-width + margin-right + right > width of the containing block```

In such cases, the CSS2 specification directs that either `left` or `right` is ignored.

How is "over-constrained" defined in CSS3? Additionally, how are CSS alignment properties and the inset-modified containing block relevant in this context?


>And it's not like it's a radically different model, it's just a slightly different way of explaining the CSS2 behavior.

 The CSS3 model is fundamentally different from CSS2, as distinct as day and night.
 
In CSS2, the inset properties define the element’s position relative to the absolute-position containing block. 
In contrast, in CSS3, these properties define the coordinates of a reference frame (e.g inset-modified containing block) relative to the absolute-position containing block and the positioned element is aligned within that frame.

For example, in CSS2, `top: 0; right: 0;` position the element at the top-right corner of the absolute-position containing block. This is aligned with established documentation, such as the following excerpt from the MDN introductory course ["Learn to style HTML using CSS"](https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Positioning#absolute_positioning):

_“Notice that the position of the element has changed. This is because `top`, `bottom`, `left`, and `right` behave differently with absolute positioning. Rather than positioning the element based on its relative position within the normal document flow, they specify the distance the element should be from each side of the containing element. For example, an absolutely positioned element set to `top: 30px` and `left: 30px` will sit 30px from the top and left edges of its containing element.”_

In CSS3, however, `top: 0; right: 0;` specify the top-right coordinates of the _inset-modified containing block_. 
The _element's position_ within this block must be inferred based on its dimensions and the remaining inset properties. This model requires additional mental steps to calculate the element’s placement, making it far less intuitive compared to the CSS2 model.
